Mathlib/MeasureTheory/PiSystem.lean

```diff
@@ -0,0 +1,767 @@
+  intro s h_s t h_t _
+  rw [Set.mem_singleton_iff.1 h_s, Set.mem_singleton_iff.1 h_t, Set.inter_self,
+    Set.mem_singleton_iff]
+#align is_pi_system.singleton IsPiSystem.singleton
+
+theorem IsPiSystem.insert_empty {α} {S : Set (Set α)} (h_pi : IsPiSystem S) :
+    IsPiSystem (insert ∅ S) := by
+  intro s hs t ht hst
+  cases' hs with hs hs
+  · simp [hs]
+  · cases' ht with ht ht
+    · simp [ht]
+    · exact Set.mem_insert_of_mem _ (h_pi s hs t ht hst)
+#align is_pi_system.insert_empty IsPiSystem.insert_empty
+
+theorem IsPiSystem.insert_univ {α} {S : Set (Set α)} (h_pi : IsPiSystem S) :
+    IsPiSystem (insert Set.univ S) := by
+  intro s hs t ht hst
+  cases' hs with hs hs
+  · cases' ht with ht ht <;> simp [hs, ht]
+  · cases' ht with ht ht
+    · simp [hs, ht]
+    · exact Set.mem_insert_of_mem _ (h_pi s hs t ht hst)
+#align is_pi_system.insert_univ IsPiSystem.insert_univ
+
+theorem IsPiSystem.comap {α β} {S : Set (Set β)} (h_pi : IsPiSystem S) (f : α → β) :
+    IsPiSystem { s : Set α | ∃ t ∈ S, f ⁻¹' t = s } := by
+  rintro _ ⟨s, hs_mem, rfl⟩ _ ⟨t, ht_mem, rfl⟩ hst
+  rw [← Set.preimage_inter] at hst⊢
+  refine' ⟨s ∩ t, h_pi s hs_mem t ht_mem _, rfl⟩
+  by_contra h
+  rw [Set.not_nonempty_iff_eq_empty] at h
+  rw [h] at hst
+  simp at hst
+#align is_pi_system.comap IsPiSystem.comap
+
+theorem isPiSystem_unionᵢ_of_directed_le {α ι} (p : ι → Set (Set α))
+    (hp_pi : ∀ n, IsPiSystem (p n)) (hp_directed : Directed (· ≤ ·) p) :
+    IsPiSystem (⋃ n, p n) := by
+  intro t1 ht1 t2 ht2 h
+  rw [Set.mem_unionᵢ] at ht1 ht2⊢
+  cases' ht1 with n ht1
+  cases' ht2 with m ht2
+  obtain ⟨k, hpnk, hpmk⟩ : ∃ k, p n ≤ p k ∧ p m ≤ p k := hp_directed n m
+  exact ⟨k, hp_pi k t1 (hpnk ht1) t2 (hpmk ht2) h⟩
+#align is_pi_system_Union_of_directed_le isPiSystem_unionᵢ_of_directed_le
+
+theorem isPiSystem_unionᵢ_of_monotone {α ι} [SemilatticeSup ι] (p : ι → Set (Set α))
+    (hp_pi : ∀ n, IsPiSystem (p n)) (hp_mono : Monotone p) : IsPiSystem (⋃ n, p n) :=
+  isPiSystem_unionᵢ_of_directed_le p hp_pi (Monotone.directed_le hp_mono)
+#align is_pi_system_Union_of_monotone isPiSystem_unionᵢ_of_monotone
+
+section Order
+
+variable {α : Type _} {ι ι' : Sort _} [LinearOrder α]
+
+theorem isPiSystem_image_Iio (s : Set α) : IsPiSystem (Iio '' s) := by
+  rintro _ ⟨a, ha, rfl⟩ _ ⟨b, hb, rfl⟩ -
+  exact ⟨a ⊓ b, inf_ind a b ha hb, Iio_inter_Iio.symm⟩
+#align is_pi_system_image_Iio isPiSystem_image_Iio
+
+theorem isPiSystem_Iio : IsPiSystem (range Iio : Set (Set α)) :=
+  @image_univ α _ Iio ▸ isPiSystem_image_Iio univ
+#align is_pi_system_Iio isPiSystem_Iio
+
+theorem isPiSystem_image_Ioi (s : Set α) : IsPiSystem (Ioi '' s) :=
+  @isPiSystem_image_Iio αᵒᵈ _ s
+#align is_pi_system_image_Ioi isPiSystem_image_Ioi
+
+theorem isPiSystem_Ioi : IsPiSystem (range Ioi : Set (Set α)) :=
+  @image_univ α _ Ioi ▸ isPiSystem_image_Ioi univ
+#align is_pi_system_Ioi isPiSystem_Ioi
+
+-- porting note: changed `max`/`min` to `⊔`/`⊓`
```

Comment:
Was there a Zulip discussion about this? If not, I started one here: https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/MeasureTheory.2EPiSystem.20!4.232160